core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -141,17 +139,24 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
     }
   }
 
-  override def getListing(): Iterable[FsApplicationHistoryInfo] = applications.values
+  override def getListing(refresh: Boolean): Iterable[FsApplicationHistoryInfo] = {
+    if (refresh) checkForLogs()
+    applications.values
+  }
 
   override def getAppUI(appId: String): Option[SparkUI] = {
     try {
-      applications.get(appId).map { info =>
+      val appOpt = applications.get(appId).orElse {
```

Comment:
I think this deserves a comment or even a separate PR. It seems you're trying to force a refresh every time the user requests an app that is not found. That can get expensive and really racy, since `getListing()` is not exactly cheap and makes assumptions about being called from a single thread.